File: src/components/app-receipt-extraction/ocr.service.ts
import {Injectable} from "../../global/di/injectable";
import { v4 as uuid } from 'uuid';

@Injectable
export class OcrService {
  private endpoint: string = 'https://thelab.cognitiveservices.azure.com';
  private key: string = '8eef6983cada4509967a088b3cef6e77';

  async detectText(url: string): Promise<AnalyzeResult> {
    const response = await fetch(`${this.endpoint}/vision/v3.1/read/analyze?language=de`,{
      method: 'POST',
      headers: new Headers({
        'Content-Type': 'application/json',
        'Ocp-Apim-Subscription-Key': this.key
      }),
      body: JSON.stringify({ url })
    });

    const resultUrl = response.headers.get('Operation-Location');

    const maxRetries = 30;
    let retries = 0;
    return await new Promise((resolve, reject) => {
      const interval = setInterval(async () => {
        const result = await fetch(resultUrl, {
          method: 'GET',
          headers: new Headers({
            'Ocp-Apim-Subscription-Key': this.key
          })
        });

        const resultJson = await result.json();

        if (resultJson.status === 'succeeded') {
          this.addIdToLines(resultJson);
          resolve(resultJson);
          clearInterval(interval);
        }

        if (resultJson.status === 'failed') {
          reject(resultJson);
          clearInterval(interval);
        }

        retries++;

        if (maxRetries <= retries) {
          reject();
          clearInterval(interval);;
        }
      }, 500);
    });
  }

  private addIdToLines(analyzeResult: AnalyzeResult) {
    analyzeResult.analyzeResult.readResults.forEach(result => {
      result.lines.forEach(line => {
        line.id = uuid();
      });
    });
  }
}

export interface AnalyzeResult {
  analyzeResult: {
    readResults: Array<ReadResult>;
  };
}

export interface ReadResult {
  page: number;
  angle: number;
  language: 'de' | 'en';
  lines: Array<Line>;
}

export interface Line {
  id: string;
  boundingBox: Array<number>;
  text: string;
}

File: backend/src/main/kotlin/ch/lucfonjallaz/drezip/bl/receipt/init/OcrService.kt
package ch.lucfonjallaz.drezip.bl.receipt.init

import ch.lucfonjallaz.drezip.bl.receipt.HttpRequestService
import org.springframework.beans.factory.annotation.Value
import org.springframework.http.*
import org.springframework.stereotype.Component

@Component
class OcrService(
        @Value("\${app.azure-blob-storage.host}") val blobStorageHost: String,
        @Value("\${app.ocr-key}") val ocrKey: String,
        @Value("\${app.ocr-url}") val ocrUrl: String,
        @Value("\${app.ocr-retry-count:50}") val ocrRetryCount: Int,
        @Value("\${app.ocr-retry-sleep-ms:500}") val ocrRetrySleepMs: Long,
        val httpRequestService: HttpRequestService
) {
    fun extractText(imageUrl: String): AzureReadResultDto? {
        val operationLocationUrl = getOperationLocationUrl("$blobStorageHost$imageUrl")

        for (i in 1..ocrRetryCount) {
            val result = getOperationResult(operationLocationUrl)

            when (result.status) {
                "succeeded" -> return result.analyzeResult?.readResults?.firstOrNull() ?: throw Exception("no text to extract found")
                "failed" -> return null
                else -> Thread.sleep(ocrRetrySleepMs)
            }
        }

        return null
    }

    private fun getOperationLocationUrl(imageUrl: String): String {
        val headers = HttpHeaders()
        headers.add("Ocp-Apim-Subscription-Key", ocrKey)
        headers.add("Content-Type", "application/json")

        val response = httpRequestService.post(ocrUrl, mapOf("url" to imageUrl), headers, Unit::class)

        return response.headers.getFirst("Operation-Location") ?: throw Exception("no operation location header present")
    }

    private fun getOperationResult(operationLocationUrl: String): AzureRequestResultsDto {
        val headers = HttpHeaders()
        headers.add("Ocp-Apim-Subscription-Key", ocrKey)

        val response = httpRequestService.get(operationLocationUrl, headers, AzureRequestResultsDto::class)

        if (response.statusCode == HttpStatus.OK) {
            return response.body ?: throw Exception("response body was empty")
        } else {
            throw Exception("OCR call failed")
        }
    }
}

data class AzureRequestResultsDto (
        val status: String,
        val analyzeResult: AzureAnalyzeResultDto?
)

data class AzureAnalyzeResultDto (
        val readResults: List<AzureReadResultDto>
)

data class AzureReadResultDto(
        val angle: Float,
        val lines: List<AzureLineDto>
)

data class AzureLineDto (
        val boundingBox: List<Int>,
        val text: String
)